src/gfx/convert.cpp

```diff
@@ -0,0 +1,831 @@
+	}
+
+	~Png() { png_destroy_read_struct(&png, &info, nullptr); }
+
+	class TilesVisitor {
+		Png const &_png;
+		bool const _columnMajor;
+		uint32_t const _width, _height;
+		uint32_t const _limit = _columnMajor ? _height : _width;
+
+	public:
+		TilesVisitor(Png const &png, bool columnMajor, uint32_t width, uint32_t height)
+			: _png(png), _columnMajor(columnMajor), _width(width), _height(height) {}
+
+		class Tile {
+			Png const &_png;
+			uint32_t const _x, _y;
+
+		public:
+			Tile(Png const &png, uint32_t x, uint32_t y) : _png(png), _x(x), _y(y) {}
+
+			uint16_t pixel(uint32_t xOfs, uint32_t yOfs) const {
+				return _png.pixel(_x + xOfs, _y + yOfs);
+			}
+		};
+
+	private:
+		struct iterator {
+			TilesVisitor const &parent;
+			uint32_t const limit;
+			uint32_t x, y;
+
+		public:
+			std::pair<uint32_t, uint32_t> coords() const { return {x, y}; }
+			Tile operator*() const { return {parent._png, x, y}; }
+
+			iterator &operator++() {
+				auto [major, minor] = parent._columnMajor ? std::tie(y, x) : std::tie(x, y);
+				major += 8;
+				if (major == limit) {
+					minor += 8;
+					major = 0;
+				}
+
+				return *this;
+			}
+
+			bool operator!=(iterator const &rhs) const {
+				return coords() != rhs.coords(); // Compare the returned coord pairs
+			}
+		};
+
+	public:
+		iterator begin() const { return {*this, _width, 0, 0}; }
+		iterator end() const {
+			iterator it{*this, _limit, _width - 8, _height - 8}; // Last valid one
+			return ++it; // Now one-past-last
+		}
+	};
+public:
+	TilesVisitor visitAsTiles(bool columnMajor) const {
+		return {*this, columnMajor, width, height};
+	}
+};
+
+class RawTiles {
+public:
+	/**
+	 * A tile which only contains indices into the image's global palette
+	 */
+	class RawTile {
+		std::array<std::array<size_t, 8>, 8> _pixelIndices{};
+
+	public:
+		// Not super clean, but it's closer to matrix notation
+		size_t &operator()(size_t x, size_t y) { return _pixelIndices[y][x]; }
+	};
+
+private:
+	std::vector<RawTile> _tiles;
+
+public:
+	/**
+	 * Creates a new raw tile, and returns a reference to it so it can be filled in
+	 */
+	RawTile &newTile() {
+		_tiles.emplace_back();
+		return _tiles.back();
+	}
+};
+
+struct AttrmapEntry {
+	size_t protoPaletteID;
+	uint16_t tileID;
+	bool yFlip;
+	bool xFlip;
+};
+
+static void outputPalettes(std::vector<Palette> const &palettes) {
+	std::filebuf output;
+	output.open(options.palettes, std::ios_base::out | std::ios_base::binary);
+
+	for (Palette const &palette : palettes) {
+		for (uint16_t color : palette) {
+			output.sputc(color & 0xFF);
+			output.sputc(color >> 8);
+		}
+	}
+}
+
+namespace unoptimized {
+
+// TODO: this is very redundant with `TileData`; try merging both?
+static void outputTileData(Png const &png, DefaultInitVec<AttrmapEntry> const &attrmap,
+                           std::vector<Palette> const &palettes,
+                           DefaultInitVec<size_t> const &mappings) {
```

Comment:
How would you feel about this indentation style for too-long function signatures, if clang-format can automate it:

```c++
static void outputTileData(
	Png const &png,
	DefaultInitVec<AttrmapEntry> const &attrmap,
	std::vector<Palette> const &palettes,
	DefaultInitVec<size_t> const &mappings
) {
```

Reply:
`BinPackArguments: false` and `BinPackParameters: false` can do this: https://stackoverflow.com/questions/29483626/